### PR TITLE
Replace recursionschemes with Matryoshka.

### PR DESCRIPTION
**Pull request by sellout**

### CODE REVIEW

**Comment by sellout** [comment]
Where is the Travis build? I want to see how this affects coverage.

**Comment by sellout** [comment]
I got Matryoshka 0.1.0 published this morning, so I'll update this with that. I might also check some of the coverage stuff.

**Comment by wemrysi** [comment]
-1.5k LOC, very nice!

**Comment by sellout** [comment]
@wemrysi Had to light a fire under your Free algebra butts 😉

File: core/src/main/scala/quasar/fp/package.scala [hunk 1]
```diff
@@ -478,4 +479,72 @@ package object fp extends TreeInstances with ListMapInstances with EitherTInstan
     */
   def pPrism[A, B](out: PartialFunction[A, B])(in: B => A) =
     Prism[A, B](out.lift)(in)
+
```
Comment (wemrysi) [inline core/src/main/scala/quasar/fp/package.scala]: Can all of the `bound*` functions become methods on `Binder`? Or if not that, functions in the companion?

File: core/src/main/scala/quasar/fp/package.scala [hunk 2]
```diff
@@ -478,4 +479,72 @@ package object fp extends TreeInstances with ListMapInstances with EitherTInstan
...
+  /**
+   Annotate (the original nodes of) a tree, by applying a function to the
+   "bound" nodes. The function is also applied to the bindings themselves
+   to determine their annotation.
+   */
+  def boundAttribute[F[_]: Functor, A](t: Fix[F])(f: Fix[F] => A)(implicit B: Binder[F]): Cofree[F, A] = {
```
Comment (wemrysi) [inline core/src/main/scala/quasar/fp/package.scala]: Can this be generalized to `boundAttribute[T[_[_]]: Recursive : Corecursive, F[_]: Functor, A]` ?

File: core/src/main/scala/quasar/fp/package.scala [hunk 3]
```diff
@@ -478,4 +479,72 @@ package object fp extends TreeInstances with ListMapInstances with EitherTInstan
...
+  def boundPara[T[_[_]]: Recursive, F[_]: Functor, A](t: T[F])(f: F[(T[F], A)] => A)(implicit B: Binder[F]): A = {
+    def loop(t: F[T[F]], b: B.G[A]): A = {
+      val newB = B.bindings(t, b)(loop(_, b))
+      B.subst(t, newB).getOrElse(f(t.map(x => (x, loop(x.project, newB)))))
+    }
+
+    loop(t.project, B.initial)
+  }
+
+  implicit def fixRenderTree[F[_]](implicit RF: RenderTree ~> λ[α => RenderTree[F[α]]]):
```
Comment (wemrysi) [inline core/src/main/scala/quasar/fp/package.scala]: Possible to move these into the `RenderTree` companion?

**Comment by wemrysi** [comment]
Had a couple comments, but other than that :+1:

File: core/src/main/scala/quasar/fp/package.scala (hunk 1 above)
Comment (sellout) [inline core/src/main/scala/quasar/fp/package.scala]: I’m not sure how they’d be methods on `Binder` … or at least how they fit my biases 😉 They’re not “injectable”.

And generally, we don’t reference things in a companion object directly. I could make a `quasar.fp.binder` package and put `Binder` and these functions in there?

I’d really like to spend some time ABTing stuff and getting rid of `Binder` all together.

Comment (wemrysi) [inline core/src/main/scala/quasar/fp/package.scala]: A `quasar.fp.binder` package would be fine. And :+1: on ABT-ing stuff, maybe we can resurrect efforts on `scala-abt`.

**Comment by wemrysi** [comment]
Updates look good, :+1:

**Pull request merged**